File: clients/go/management.go
package agenticdriver

import (
	"context"
	"regexp"
	"strings"
)

// ProviderConfiguration contains secret references, never resolved secret values.
type ProviderConfiguration struct {
	ID        string `json:"id"`
	Kind      string `json:"kind"`
	AccountID string `json:"accountId,omitempty"`
	Name      string `json:"name,omitempty"`
	Enabled   *bool  `json:"enabled,omitempty"`
	// nil permits every explicitly selected model; a pointer to an empty slice denies all.
	Models           *[]string           `json:"models,omitempty"`
	APIKeyRef        map[string]any      `json:"apiKeyRef,omitempty"`
	BaseURL          string              `json:"baseUrl,omitempty"`
	Binary           string              `json:"binary,omitempty"`
	AccountDirectory string              `json:"accountDirectory,omitempty"`
	ReasoningEffort  string              `json:"reasoningEffort,omitempty"`
	ApplicationTools string              `json:"applicationTools,omitempty"`
	InputMediaTypes  map[string][]string `json:"inputMediaTypes,omitempty"`
	ExtensionID      string              `json:"extensionId,omitempty"`
	ExtensionVersion string              `json:"extensionVersion,omitempty"`
	Settings         map[string]any      `json:"settings,omitempty"`
	SecretRefs       map[string]any      `json:"secretRefs,omitempty"`
}

// ProviderConnectionMethod describes setup only; it does not grant execution.
type ProviderConnectionMethod struct {
	ID              string `json:"id"`
	Label           string `json:"label"`
	Description     string `json:"description"`
	Interaction     string `json:"interaction"`
	CredentialOwner string `json:"credentialOwner"`
}
type ProviderDefinition struct {
	Kind         string                     `json:"kind"`
	Name         string                     `json:"name"`
	Description  string                     `json:"description"`
	Category     string                     `json:"category"`
	Protocol     string                     `json:"protocol"`
	Methods      []ProviderConnectionMethod `json:"methods"`
	Requirements string                     `json:"requirements,omitempty"`
	DocsURL      string                     `json:"docsUrl,omitempty"`
}
type ManagementSnapshot struct {
	Version             int                     `json:"version"`
	Revision            string                  `json:"revision"`
	Providers           []ProviderConfiguration `json:"providers"`
	SupportedKinds      []string                `json:"supportedKinds"`
	ProviderDefinitions *[]ProviderDefinition   `json:"providerDefinitions,omitempty"`
	ExecutionProviders  *[]string               `json:"executionProviders,omitempty"`
}
type ConfigureProvider struct {
	Revision string                `json:"revision"`
	Provider ProviderConfiguration `json:"provider"`
	// Write only. Never returned by the host.
	APIKey string `json:"apiKey,omitempty"`
}

var managementRevision = regexp.MustCompile(`^[a-f0-9]{64}$`)
var connectionMethodID = regexp.MustCompile(`^[a-z][a-z0-9-]{0,79}$`)

func validProviderDefinition(d ProviderDefinition) bool {
	text := func(s string, max int) bool { return len(s) > 0 && len(s) <= max }
	if !text(d.Kind, 80) || !text(d.Name, 100) || !text(d.Description, 1000) || !text(d.Protocol, 100) || len(d.Methods) < 1 || len(d.Methods) > 8 || len(d.Requirements) > 2000 || (d.DocsURL != "" && (!strings.HasPrefix(d.DocsURL, "https://") || len(d.DocsURL) > 2000)) {
		return false
	}
	switch d.Category {
	case "native", "api", "compatible", "fixture":
	default:
		return false
	}
	for _, m := range d.Methods {
		if !connectionMethodID.MatchString(m.ID) || !text(m.Label, 100) || !text(m.Description, 1000) {
			return false
		}
		if !connectionMethodID.MatchString(m.Interaction) {
			return false
		}
		switch m.CredentialOwner {
		case "native-runtime", "host", "none":
		default:
			return false
		}
	}
	return true
}

func validManagement(value ManagementSnapshot, id string) bool {
	if value.Version != 1 || !managementRevision.MatchString(value.Revision) || value.Providers == nil || len(value.Providers) > 32 || value.SupportedKinds == nil {
		return false
	}
	if value.ProviderDefinitions != nil {
		if len(*value.ProviderDefinitions) > 32 {
			return false
		}
		for _, definition := range *value.ProviderDefinitions {
			if !validProviderDefinition(definition) {
				return false
			}
		}
	}
	seen := map[string]bool{}
	for _, p := range value.Providers {
		if p.ID == "" || p.Kind == "" || seen[p.ID] {
			return false
		}
		seen[p.ID] = true
	}
	return id == "" || seen[id]
}
func (c *Client) managementRequest(ctx context.Context, path string, body any, id string) (ManagementSnapshot, error) {
	var value ManagementSnapshot
	res, err := c.request(ctx, path, body, false)
	if err != nil {
		return value, err
	}
	if err = decode(res, &value); err != nil {
		return value, err
	}
	if !validManagement(value, id) {
		return value, &Error{Code: "INVALID_RESPONSE", Message: "The host returned invalid or mismatched provider settings. Refresh before retrying."}
	}
	return value, nil
}
func (c *Client) Management(ctx context.Context) (ManagementSnapshot, error) {
	return c.managementRequest(ctx, "v1/management", nil, "")
}
func (c *Client) ConfigureProvider(ctx context.Context, input ConfigureProvider) (ManagementSnapshot, error) {
	return c.managementRequest(ctx, "v1/management/providers", input, input.Provider.ID)
}

File: clients/rust/src/management.rs
//! Remote host administration, independently granted from execution access.
use crate::{Error, Result};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;

#[derive(Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ProviderConfiguration {
    pub id: String,
    pub kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub account_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enabled: Option<bool>,
    /// None permits every explicit model; Some(vec![]) denies all.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub models: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_key_ref: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub base_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub binary: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub account_directory: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reasoning_effort: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub application_tools: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input_media_types: Option<BTreeMap<String, Vec<String>>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extension_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extension_version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub settings: Option<BTreeMap<String, Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub secret_refs: Option<BTreeMap<String, Value>>,
}
/// Setup metadata only, not account availability or an execution grant.
#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderConnectionMethod {
    pub id: String,
    pub label: String,
    pub description: String,
    pub interaction: String,
    pub credential_owner: String,
}
#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderDefinition {
    pub kind: String,
    pub name: String,
    pub description: String,
    pub category: String,
    pub protocol: String,
    pub methods: Vec<ProviderConnectionMethod>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub requirements: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub docs_url: Option<String>,
}
#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManagementSnapshot {
    pub version: u32,
    pub revision: String,
    pub providers: Vec<ProviderConfiguration>,
    pub supported_kinds: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider_definitions: Option<Vec<ProviderDefinition>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub execution_providers: Option<Vec<String>>,
}
#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConfigureProvider {
    pub revision: String,
    pub provider: ProviderConfiguration,
    /// Write only. Omit when keeping the existing credential reference.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_key: Option<String>,
}
fn valid_definition(d: &ProviderDefinition) -> bool {
    let text = |s: &str, max: usize| !s.is_empty() && s.len() <= max;
    text(&d.kind, 80)
        && text(&d.name, 100)
        && text(&d.description, 1000)
        && text(&d.protocol, 100)
        && matches!(
            d.category.as_str(),
            "native" | "api" | "compatible" | "fixture"
        )
        && (1..=8).contains(&d.methods.len())
        && d.requirements.as_ref().is_none_or(|s| text(s, 2000))
        && d.docs_url
            .as_ref()
            .is_none_or(|s| text(s, 2000) && s.starts_with("https://"))
        && d.methods.iter().all(|m| {
            text(&m.id, 80)
                && m.id.as_bytes()[0].is_ascii_lowercase()
                && m.id
                    .bytes()
                    .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == b'-')
                && text(&m.label, 100)
                && text(&m.description, 1000)
                && text(&m.interaction, 80)
                && m.interaction.as_bytes()[0].is_ascii_lowercase()
                && m.interaction
                    .bytes()
                    .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == b'-')
                && matches!(
                    m.credential_owner.as_str(),
                    "native-runtime" | "host" | "none"
                )
        })
}
pub(crate) fn snapshot(value: Value, id: Option<&str>) -> Result<ManagementSnapshot> {
    let parsed: ManagementSnapshot = serde_json::from_value(value)
        .map_err(|_| Error::Protocol("Invalid provider settings response."))?;
    let mut ids = std::collections::HashSet::new();
    if parsed.version != 1
        || parsed.revision.len() != 64
        || !parsed
            .revision
            .bytes()
            .all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b))
        || parsed.providers.len() > 32
        || parsed.supported_kinds.len() > 32
        || parsed
            .provider_definitions
            .as_ref()
            .is_some_and(|definitions| {
                definitions.len() > 32 || definitions.iter().any(|d| !valid_definition(d))
            })
        || parsed
            .providers
            .iter()
            .any(|p| p.id.is_empty() || p.kind.is_empty() || !ids.insert(p.id.as_str()))
        || id.is_some_and(|id| !ids.contains(id))
    {
        return Err(Error::Protocol(
            "Invalid or mismatched provider settings. Refresh before retrying.",
        ));
    }
    Ok(parsed)
}
